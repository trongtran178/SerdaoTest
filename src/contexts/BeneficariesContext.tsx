import AsyncStorage from '@react-native-async-storage/async-storage';
import debounce from 'lodash/debounce';
import toString from 'lodash/toString';
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState,
} from 'react';
import {Beneficiary} from '../models';
import {notifyMessage} from '../utils';
import { ECacheKeys } from '../constants';

export type BeneficariesContextType = {
  beneficiaries: Beneficiary[];
  addBeneficiary: (beneficary: unknown) => Beneficiary | undefined;
};

const BeneficariesContext = createContext<BeneficariesContextType | undefined>(
  undefined,
);

export const useBeneficiaries = () => useContext(BeneficariesContext);

export const BeneficiariesProvider = ({
  children,
}: {
  children: React.ReactNode | React.ReactNode[];
}) => {
  const [beneficiaries, setBeneficiaries] = useState<Beneficiary[]>([]);

  useEffect(() => {}, []);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const saveDataToAsyncStorage = useCallback(
    debounce(async () => {
      try {
        await AsyncStorage.setItem(
          ECacheKeys.BENEFICIARIES_CONTEXT,
          JSON.stringify(beneficiaries),
        );
      } catch (e) { }
    }, 1000),
    [beneficiaries],
  );

  useEffect(() => {
    saveDataToAsyncStorage();
  }, [beneficiaries, saveDataToAsyncStorage]);

  useEffect(() => {
    const getBeneficiaries = async () => {
      const result = await AsyncStorage.getItem(ECacheKeys.BENEFICIARIES_CONTEXT);
      if (result) {
        const parsed = JSON.parse(result);
        setBeneficiaries(parsed);
      }
    };
    getBeneficiaries();
  }, []);

  const addBeneficiary = (beneficary: unknown) => {
    try {
      const newBeneficiary = Beneficiary.instantiate(beneficary);
      setBeneficiaries(prevBeneficiaries => [
        ...prevBeneficiaries,
        newBeneficiary,
      ]);
      notifyMessage('Beneficiary added successfully');
      return newBeneficiary;
    } catch (e) {
      notifyMessage(toString(e));
    }
  };
  return (
    <BeneficariesContext.Provider value={{beneficiaries, addBeneficiary}}>
      {children}
    </BeneficariesContext.Provider>
  );
};
