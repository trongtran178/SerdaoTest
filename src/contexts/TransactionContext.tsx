import React, {
  createContext,
  useState,
  useContext,
  useEffect,
  useCallback,
} from 'react';
import {Transaction, Account} from '../models';
import get from 'lodash/get';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {debounce, isNumber} from 'lodash';
import { ECacheKeys } from '../constants';
export type TransactionContextType = {
  transactions: Transaction[];
  addTransaction: (amount: string, account: unknown) => void;
  balance: number;
};
const TransactionContext = createContext<TransactionContextType | undefined>(
  undefined,
);

export const useTransactions = () => useContext(TransactionContext);

export const TransactionProvider = ({
  children,
}: {
  children: React.ReactNode | React.ReactNode[];
}) => {
  const [transactions, setTransactions] = useState<Transaction[]>([]);
  const [balance, setBalance] = useState<number>(1000);

  const addTransaction = (amount: string, account: unknown) => {
    const accountTransaction = Account.instantiate({
      name: get(account, 'name', ''),
      iban: get(account, 'iban', ''),
    });
    const regex = /^-?\d+(\.\d+)?$/;
    
    if(/^-?\d+(\.\d+)?$/.test(amount) === false)
      throw new Error('Amount must be a number');
    const amountParsed = parseFloat(amount);
    const newTransaction: Transaction = Transaction.instantiate({
      id: Date.now(),
      amount: amountParsed,
      account: accountTransaction,
      currentBalance: balance,
    });
    setTransactions(prevTransactions => [...prevTransactions, newTransaction]);
    setBalance(prevBalance => prevBalance - parseFloat(amount));
  };

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const saveDataToAsyncStorage = useCallback(
    debounce(() => {
      try {
        AsyncStorage.setItem(
          ECacheKeys.TRANSACTIONS_CONTEXT,
          JSON.stringify({transactions, balance}),
        );
      } catch (e) {}
    }, 1000),
    [transactions, balance],
  );

  useEffect(() => {
    saveDataToAsyncStorage();
  }, [transactions, balance, saveDataToAsyncStorage]);

  useEffect(() => {
    const getTransactions = async () => {
      try {
        const data = await AsyncStorage.getItem(ECacheKeys.TRANSACTIONS_CONTEXT);
        if (data) {
          const {transactions: transactionsCached, balance: balanceCached} =
            JSON.parse(data);
          setTransactions(transactionsCached);
          setBalance(balanceCached);
        }
      } catch (e) {}
    };

    getTransactions();
  }, []);

  return (
    <TransactionContext.Provider
      value={{transactions, addTransaction, balance}}>
      {children}
    </TransactionContext.Provider>
  );
};
