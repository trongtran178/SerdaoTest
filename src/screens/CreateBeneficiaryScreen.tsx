import isNil from 'lodash/isNil';
import React, {useCallback, useState} from 'react';
import {Image, StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import {CommonTextInput} from '../components/CommonTextInput';
import {useBeneficiaries} from '../contexts/BeneficariesContext';
import {ParamListBase} from '@react-navigation/native';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {Icons} from '../assets';

type Props = NativeStackScreenProps<ParamListBase, 'CreateBeneficiary'>;

const CreateBeneficaryScreen = ({navigation}: Props) => {
  const [firstName, setFirstName] = useState<string>('');
  const [lastName, setLastName] = useState<string>('');
  const [iban, setIban] = useState<string>('');

  const beneficiariesContext = useBeneficiaries();

  if (isNil(beneficiariesContext)) {
    return null;
  }
  const {addBeneficiary} = beneficiariesContext;

  // eslint-disable-next-line react-hooks/rules-of-hooks
  const onCreate = useCallback(() => {
    const newBeneficiary = addBeneficiary({firstName, lastName, iban});
    if (!isNil(newBeneficiary)) {
      setFirstName('');
      setLastName('');
      setIban('');
      navigation.goBack();
    }
  }, [addBeneficiary, firstName, iban, lastName, navigation]);

  return (
    <View style={styles.container}>
      <View style={styles.inputContainer}>
        <CommonTextInput
          testID='textinput_firstname'
          onChangeText={setFirstName}
          value={firstName}
          maxLength={50}
          placeholder="Enter first name"
        />
        <View>
          <Image source={Icons.name} style={styles.icon} />
        </View>
      </View>
      <View style={styles.inputContainer}>
        <CommonTextInput
          testID='textinput_lastname'
          onChangeText={setLastName}
          value={lastName}
          maxLength={50}
          placeholder="Enter last name"
        />
        <View>
          <Image source={Icons.name} style={styles.icon} />
        </View>
      </View>

      <View style={styles.inputContainer}>
        <CommonTextInput
          testID='textinput_iban'
          onChangeText={setIban}
          value={iban}
          maxLength={34}
          placeholder="Enter IBAN number"
        />
        <View>
          <Image source={Icons.iban_number} style={styles.icon} />
        </View>
      </View>

      <TouchableOpacity style={styles.button} onPress={onCreate}>
        <Text style={styles.buttonText}>Create beneficiary</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textInput: {
    height: 40,
    borderColor: 'gray',
    borderWidth: 1,
    width: '80%',
    marginVertical: 8,
  },
  button: {
    backgroundColor: '#4CAF50', // Material Design green
    marginVertical: 12,
    padding: 12,
    borderRadius: 12,
    alignItems: 'center',
    elevation: 2, // for Android shadow
    shadowColor: '#000', // for iOS shadow
    shadowOffset: {width: 0, height: 2}, // for iOS shadow
    shadowOpacity: 0.1, // for iOS shadow
    shadowRadius: 5, // for iOS shadow
  },
  buttonText: {
    color: '#fff',
    fontSize: 16,
    fontWeight: 'bold',
  },
  icon: {
    position: 'absolute',
    top: -12,
    right: 12,
    width: 20,
    height: 20,
  },
  inputContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    borderRadius: 5,
  },
});

export default CreateBeneficaryScreen;
