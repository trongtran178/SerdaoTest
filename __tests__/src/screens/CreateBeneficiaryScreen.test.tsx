import 'react-native';
import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import CreateBeneficiaryScreen from '../../../src/screens/CreateBeneficiaryScreen';
import { describe, it, jest, expect } from '@jest/globals';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { ParamListBase } from '@react-navigation/native';
import { BeneficariesContextType, BeneficiariesProvider, useBeneficiaries } from '../../../src/contexts/BeneficariesContext';
import { Command } from 'commander';
import { Text } from 'react-native';

describe('CreateBeneficiaryScreen', () => {
    const navigationMock = {
        navigate: jest.fn(),
        goBack: jest.fn(),
        canGoBack: jest.fn(),
        setOptions: jest.fn(),
        setParams: jest.fn(),
        dispatch: jest.fn(),
        isFocused: jest.fn(),
        dangerouslyGetState: jest.fn(),
        dangerouslyGetParent: jest.fn(),
        getState: jest.fn(),
        reset: jest.fn(),
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addKeyedListener: jest.fn(),
        removeKeyedListener: jest.fn(),
        launch: jest.fn(),
        dismiss: jest.fn(),
        getParent: jest.fn(),
        getKey: jest.fn().mockReturnValue("Home"),
        isFirstRouteInParent: jest.fn().mockReturnValue(true),
        type: "native",
        name: "Home",
        replace: jest.fn(),
        push: jest.fn(),
        pop: jest.fn(),
        popToTop: jest.fn(),
    } as unknown as NativeStackNavigationProp<ParamListBase, "CreateBeneficiary", undefined>;
    
    const routeMock = {
        key: "CreateBeneficiary",
        name: "CreateBeneficiary",
        params: {}
      } as const;

  it('renders correctly', () => {
    const routeMock = {
        key: "CreateBeneficiary",
        name: "CreateBeneficiary",
        params: {}
      } as const;
  

    const { toJSON } = render(
        <BeneficiariesProvider>
            <CreateBeneficiaryScreen navigation={navigationMock} route={routeMock} />
        </BeneficiariesProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should return null when beneficiaries context is nil', () => {
 
    const routeMock = {
      key: "CreateBeneficiary",
      name: "CreateBeneficiary",
      params: {}
    } as const;
  

    const { toJSON } = render(
      <CreateBeneficiaryScreen navigation={navigationMock} route={routeMock} />
    );

    expect(toJSON()).toBeNull();
  });

  
  it('should call onCreate when press Create beneficiary', () => {  

    const TestComponent = () => { 
        const beneficiariesContext = useBeneficiaries();
        const {beneficiaries} = beneficiariesContext as BeneficariesContextType;
        return (
            <> 
                {beneficiaries.length > 0 && <Text>{beneficiaries[0].name}</Text>}
                <CreateBeneficiaryScreen navigation={navigationMock} route={routeMock} />
            </>
        )
    }

    const { getByText, getByTestId } = render(
      <BeneficiariesProvider>
        <TestComponent />
      </BeneficiariesProvider>
    );
    fireEvent.changeText(getByTestId('textinput_firstname'), 'John');
    fireEvent.changeText(getByTestId('textinput_lastname'), 'Doe');
    fireEvent.changeText(getByTestId('textinput_iban'), 'AT483200000012345864');
    fireEvent.press(getByText('Create beneficiary'));

    expect(getByText('John Doe')).toBeDefined();

    fireEvent.changeText(getByTestId('textinput_firstname'), '');
    fireEvent.changeText(getByTestId('textinput_lastname'), '');
    fireEvent.changeText(getByTestId('textinput_iban'), 'AT483200000012345864');
    fireEvent.press(getByText('Create beneficiary'));
    expect(getByTestId('textinput_iban').props.value).toEqual('AT483200000012345864');
  });

});
