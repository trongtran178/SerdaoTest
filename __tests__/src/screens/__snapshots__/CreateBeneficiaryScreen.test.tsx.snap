// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBeneficiaryScreen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 5,
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      maxLength={50}
      onChangeText={[Function]}
      placeholder="Enter first name"
      style={
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
          "borderRadius": 10,
          "borderWidth": 1,
          "elevation": 2,
          "height": 50,
          "marginBottom": 15,
          "marginVertical": 8,
          "paddingHorizontal": 10,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 5,
          "width": "80%",
        }
      }
      testID="textinput_firstname"
      value=""
    />
    <View>
      <Image
        source={
          {
            "uri": "https://cdn-icons-png.flaticon.com/512/149/149071.png",
          }
        }
        style={
          {
            "height": 20,
            "position": "absolute",
            "right": 12,
            "top": -12,
            "width": 20,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 5,
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      maxLength={50}
      onChangeText={[Function]}
      placeholder="Enter last name"
      style={
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
          "borderRadius": 10,
          "borderWidth": 1,
          "elevation": 2,
          "height": 50,
          "marginBottom": 15,
          "marginVertical": 8,
          "paddingHorizontal": 10,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 5,
          "width": "80%",
        }
      }
      testID="textinput_lastname"
      value=""
    />
    <View>
      <Image
        source={
          {
            "uri": "https://cdn-icons-png.flaticon.com/512/149/149071.png",
          }
        }
        style={
          {
            "height": 20,
            "position": "absolute",
            "right": 12,
            "top": -12,
            "width": 20,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 5,
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      maxLength={34}
      onChangeText={[Function]}
      placeholder="Enter IBAN number"
      style={
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
          "borderRadius": 10,
          "borderWidth": 1,
          "elevation": 2,
          "height": 50,
          "marginBottom": 15,
          "marginVertical": 8,
          "paddingHorizontal": 10,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 5,
          "width": "80%",
        }
      }
      testID="textinput_iban"
      value=""
    />
    <View>
      <Image
        source={
          {
            "uri": "https://cdn-icons-png.flaticon.com/512/4930/4930340.png",
          }
        }
        style={
          {
            "height": 20,
            "position": "absolute",
            "right": 12,
            "top": -12,
            "width": 20,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#4CAF50",
        "borderRadius": 12,
        "elevation": 2,
        "marginVertical": 12,
        "opacity": 1,
        "padding": 12,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 5,
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Create beneficiary
    </Text>
  </View>
</View>
`;
